## Описание проекта

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.
Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Вам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком.
Постройте модель с предельно большим значением F1-меры. Чтобы сдать проект успешно, нужно довести метрику до 0.59. Проверьте F1-меру на тестовой выборке самостоятельно.
Дополнительно измеряйте AUC-ROC, сравнивайте её значение с F1-мерой.

## Описание данных

Признаки

    RowNumber — индекс строки в данных
    CustomerId — уникальный идентификатор клиента
    Surname — фамилия
    CreditScore — кредитный рейтинг
    Geography — страна проживания
    Gender — пол
    Age — возраст
    Tenure — сколько лет человек является клиентом банка
    Balance — баланс на счёте
    NumOfProducts — количество продуктов банка, используемых клиентом
    HasCrCard — наличие кредитной карты
    IsActiveMember — активность клиента
    EstimatedSalary — предполагаемая зарплата

Целевой признак

    Exited — факт ухода клиента

In [18]:
import pandas as pd
import numpy as np
from sklearn.utils import shuffle
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score

In [9]:
# загрузим и посмотрим на данные
data = pd.read_csv('https://code.s3.yandex.net/datasets/Churn.csv')
data.name = 'churn_clients'
data.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [15]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


In [7]:
data.describe().T

count          mean           std          min  \
RowNumber        10000.0  5.000500e+03   2886.895680         1.00   
CustomerId       10000.0  1.569094e+07  71936.186123  15565701.00   
CreditScore      10000.0  6.505288e+02     96.653299       350.00   
Age              10000.0  3.892180e+01     10.487806        18.00   
Tenure            9091.0  4.997690e+00      2.894723         0.00   
Balance          10000.0  7.648589e+04  62397.405202         0.00   
NumOfProducts    10000.0  1.530200e+00      0.581654         1.00   
HasCrCard        10000.0  7.055000e-01      0.455840         0.00   
IsActiveMember   10000.0  5.151000e-01      0.499797         0.00   
EstimatedSalary  10000.0  1.000902e+05  57510.492818        11.58   
Exited           10000.0  2.037000e-01      0.402769         0.00   

                         25%           50%           75%          max  
RowNumber            2500.75  5.000500e+03  7.500250e+03     10000.00  
CustomerId       15628528.25  1.569074e+07  1.575323e+07  15815690.00  
CreditScore           584.00  6.520000e+02  7.180000e+02       850.00  
Age                    32.00  3.700000e+01  4.400000e+01        92.00  
Tenure                  2.00  5.000000e+00  7.000000e+00        10.00  
Balance                 0.00  9.719854e+04  1.276442e+05    250898.09  
NumOfProducts           1.00  1.000000e+00  2.000000e+00         4.00  
HasCrCard               0.00  1.000000e+00  1.000000e+00         1.00  
IsActiveMember          0.00  1.000000e+00  1.000000e+00         1.00  
EstimatedSalary     51002.11  1.001939e+05  1.493882e+05    199992.48  
Exited                  0.00  0.000000e+00  0.000000e+00         1.00

In [13]:
# проверим датасет на пропуски
def get_missing_values(data: pd.DataFrame) -> None:
    """
    Выводит данные о пропусках в колонках по датафрейму.
    Не изменяет данные внутри датафрейма.

    :param data: pd.DataFrame
    :return: None
    """
    # получаем имена колонок датафрейма
    columns = data.columns.to_list()
    data_len = len(data)
    # объявляем счетчик
    counter = -1
    print('='*60)
    # если есть пропуски в данных - выводим информацию о пропусках по колонкам
    if sum(data.isnull().sum()) > 0:
        print(f'Количество записей в датафрейме {data.name}: {data_len} \n')
        print(f'В датафрейме {data.name} имеются следующие пропуски:')
        for i in data.isnull().sum():
            counter += 1
            if i > 0:
                print(f'  - в колонке {columns[counter]}: {i} пропусков, это {i/data_len:0.2%} об общего объема данных')
    else:
        print(f'Отлично, в датафрейме {data.name} отсутствуют пропуски.')

# посмотрим на пропуски в данных
get_missing_values(data)

Количество записей в датафрейме churn_clients: 10000 

В датафрейме churn_clients имеются следующие пропуски:
  - в колонке Tenure: 909 пропусков, это 9.09% об общего объема данных


Имеем 909 пропусков в колонке Tenure, в которой содержится информация о том, сколько лет человек является клиентом банка.

In [32]:
data.head(30)

RowNumber  CustomerId     Surname  CreditScore Geography  Gender  Age  \
0           1    15634602    Hargrave          619    France  Female   42   
1           2    15647311        Hill          608     Spain  Female   41   
2           3    15619304        Onio          502    France  Female   42   
3           4    15701354        Boni          699    France  Female   39   
4           5    15737888    Mitchell          850     Spain  Female   43   
5           6    15574012         Chu          645     Spain    Male   44   
6           7    15592531    Bartlett          822    France    Male   50   
7           8    15656148      Obinna          376   Germany  Female   29   
8           9    15792365          He          501    France    Male   44   
9          10    15592389          H?          684    France    Male   27   
10         11    15767821      Bearce          528    France    Male   31   
11         12    15737173     Andrews          497     Spain    Male   24   
12         13    15632264         Kay          476    France  Female   34   
13         14    15691483        Chin          549    France  Female   25   
14         15    15600882       Scott          635     Spain  Female   35   
15         16    15643966     Goforth          616   Germany    Male   45   
16         17    15737452       Romeo          653   Germany    Male   58   
17         18    15788218   Henderson          549     Spain  Female   24   
18         19    15661507     Muldrow          587     Spain    Male   45   
19         20    15568982         Hao          726    France  Female   24   
20         21    15577657    McDonald          732    France    Male   41   
21         22    15597945    Dellucci          636     Spain  Female   32   
22         23    15699309   Gerasimov          510     Spain  Female   38   
23         24    15725737      Mosman          669    France    Male   46   
24         25    15625047         Yen          846    France  Female   38   
25         26    15738191     Maclean          577    France    Male   25   
26         27    15736816       Young          756   Germany    Male   36   
27         28    15700772     Nebechi          571    France    Male   44   
28         29    15728693  McWilliams          574   Germany  Female   43   
29         30    15656300    Lucciano          411    France    Male   29   

    Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0      2.0       0.00              1          1               1   
1      1.0   83807.86              1          0               1   
2      8.0  159660.80              3          1               0   
3      1.0       0.00              2          0               0   
4      2.0  125510.82              1          1               1   
5      8.0  113755.78              2          1               0   
6      7.0       0.00              2          1               1   
7      4.0  115046.74              4          1               0   
8      4.0  142051.07              2          0               1   
9      2.0  134603.88              1          1               1   
10     6.0  102016.72              2          0               0   
11     3.0       0.00              2          1               0   
12    10.0       0.00              2          1               0   
13     5.0       0.00              2          0               0   
14     7.0       0.00              2          1               1   
15     3.0  143129.41              2          0               1   
16     1.0  132602.88              1          1               0   
17     9.0       0.00              2          1               1   
18     6.0       0.00              1          0               0   
19     6.0       0.00              2          1               1   
20     8.0       0.00              2          1               1   
21     8.0       0.00              2          1               0   
22     4.0       0.00              1          1               0   
23     3

In [33]:
data[data['Tenure'].isna()]

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
30           31    15589475    Azikiwe          591     Spain  Female   39   
48           49    15766205        Yin          550   Germany    Male   38   
51           52    15768193  Trevisani          585   Germany    Male   36   
53           54    15702298   Parkhill          655   Germany    Male   41   
60           61    15651280     Hunter          742   Germany    Male   35   
...         ...         ...        ...          ...       ...     ...  ...   
9944       9945    15703923    Cameron          744   Germany    Male   41   
9956       9957    15707861      Nucci          520    France  Female   46   
9964       9965    15642785    Douglas          479    France    Male   34   
9985       9986    15586914     Nepean          659    France    Male   36   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
30       NaN       0.00              3          1               0   
48       NaN  103391.38              1          0               1   
51       NaN  146050.97              2          0               0   
53       NaN  125561.97              1          0               0   
60       NaN  136857.00              1          0               0   
...      ...        ...            ...        ...             ...   
9944     NaN  190409.34              2          1               1   
9956     NaN   85216.61              1          1               0   
9964     NaN  117593.48              2          0               0   
9985     NaN  123841.49              2          1               0   
9999     NaN  130142.79              1          1               0   

      EstimatedSalary  Exited  
30          140469.38       1  
48           90878.13       0  
51           86424.57       0  
53          164040.94       1  
60           84509.57       0  
...               ...     ...  
9944        138361.48       0  
9956        117369.52       1  
9964        113308.29       0  
9985         96833.00       0  
9999         38190.78       0  

[909 rows x 14 columns]

In [21]:
np.ceil(pd.pivot_table(data[data['Tenure'].notna()], index=['Geography', 'Age', 'Gender'], values='Tenure').transform('mean'))

ValueError: Function did not transform

In [34]:
data['Tenure'] = data['Tenure'].fillna(np.ceil(data.groupby(by=['Geography', 'Age', 'Gender'])['Tenure'].transform('mean')))

In [37]:
data.head(40)

RowNumber  CustomerId        Surname  CreditScore Geography  Gender  Age  \
0           1    15634602       Hargrave          619    France  Female   42   
1           2    15647311           Hill          608     Spain  Female   41   
2           3    15619304           Onio          502    France  Female   42   
3           4    15701354           Boni          699    France  Female   39   
4           5    15737888       Mitchell          850     Spain  Female   43   
5           6    15574012            Chu          645     Spain    Male   44   
6           7    15592531       Bartlett          822    France    Male   50   
7           8    15656148         Obinna          376   Germany  Female   29   
8           9    15792365             He          501    France    Male   44   
9          10    15592389             H?          684    France    Male   27   
10         11    15767821         Bearce          528    France    Male   31   
11         12    15737173        Andrews          497     Spain    Male   24   
12         13    15632264            Kay          476    France  Female   34   
13         14    15691483           Chin          549    France  Female   25   
14         15    15600882          Scott          635     Spain  Female   35   
15         16    15643966        Goforth          616   Germany    Male   45   
16         17    15737452          Romeo          653   Germany    Male   58   
17         18    15788218      Henderson          549     Spain  Female   24   
18         19    15661507        Muldrow          587     Spain    Male   45   
19         20    15568982            Hao          726    France  Female   24   
20         21    15577657       McDonald          732    France    Male   41   
21         22    15597945       Dellucci          636     Spain  Female   32   
22         23    15699309      Gerasimov          510     Spain  Female   38   
23         24    15725737         Mosman          669    France    Male   46   
24         25    15625047            Yen          846    France  Female   38   
25         26    15738191        Maclean          577    France    Male   25   
26         27    15736816          Young          756   Germany    Male   36   
27         28    15700772        Nebechi          571    France    Male   44   
28         29    15728693     McWilliams          574   Germany  Female   43   
29         30    15656300       Lucciano          411    France    Male   29   
30         31    15589475        Azikiwe          591     Spain  Female   39   
31         32    15706552  Odinakachukwu          533    France    Male   36   
32         33    15750181      Sanderson          553   Germany    Male   41   
33         34    15659428        Maggard          520     Spain  Female   42   
34         35    15732963       Clements          722     Spain  Female   29   
35         36    15794171       Lombardo          475    France  Female   45   
36         37    15788448         Watson          490     Spain    Male   31   
37         38    15729599        Lorenzo          804     Spain    Male   33   
38         39    15717426      Armstrong          850    France    Male   36   
39         40    15585768        Cameron          582   Germany    Male   41   

       Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0    2.000000       0.00              1          1               1   
1    1.000000   83807.86              1          0               1   
2    8.000000  159660.80              3          1               0   
3    1.000000       0.00              2          0               0   
4    2.000000  125510.82              1          1               1   
5    8.000000  113755.78              2          1               0   
6    7.000000       0.00              2          1               1   
7    4.000000  115046.74              4          1               0   
8    4.000000  142051.07              2          0               1   
9    2.000000  134603.8

In [ ]:
def upsample(features, target, repeat):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)

    features_upsampled, target_upsampled = shuffle(
        features_upsampled, target_upsampled, random_state=12345)

    return features_upsampled, target_upsampled